File: built-in plugins dsp/mb-master-deesser.cpp
/**
 * MB Master De-Esser
 * Category : effect
 * Type     : mastering
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Broadband de-esser optimized for mix bus sibilance control
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MASTER_DEESSER_H
#define MB_MASTER_DEESSER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMasterDeesser : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-master-deesser";
    static constexpr const char* PLUGIN_NAME    = "MB Master De-Esser";
    static constexpr const char* PLUGIN_TYPE    = "mastering";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float frequency = 6000f;  // range [2000, 12000]
    float threshold = -20f;  // range [-40, 0]
    float range = -6f;  // range [-24, 0]
    float mix = 1f;  // range [0, 1]
    };

    MbMasterDeesser() = default;
    ~MbMasterDeesser() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.frequency = std::clamp(params.frequency, 2000f, 12000f);
        params.threshold = std::clamp(params.threshold, -40f, 0f);
        params.range = std::clamp(params.range, -24f, 0f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Master De-Esser
        return input;
    }
};

#endif // MB_MASTER_DEESSER_H

File: built-in plugins dsp/mb-master-deesser.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMasterDeesserPlugin: PluginDefinition = {
  id: "mb-master-deesser",
  slug: "mb-master-deesser",
  name: "MB Master De-Esser",
  category: "effect",
  type: "mastering" as any,
  version: "1.0.0",
  description: "Broadband de-esser optimized for mix bus sibilance control",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "frequency",
      name: "Frequency",
      type: "float",
      defaultValue: 6000,
      minValue: 2000,
      maxValue: 12000,
      automatable: true,
    },
    {
      id: "threshold",
      name: "Threshold",
      type: "float",
      defaultValue: -20,
      minValue: -40,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "range",
      name: "Range",
      type: "float",
      defaultValue: -6,
      minValue: -24,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { frequency: 6000, threshold: -20, range: -6, mix: 1 },
};

export default MbMasterDeesserPlugin;
